Src/WindowsServer/WindowsServer.Shared/AppServicesHeartbeatTelemetryModule.cs

```diff
@@ -0,0 +1,161 @@
+    /// are specific to Azure App Services (Web Apps, Functions, etc...).
+    /// </summary>
+    public sealed class AppServicesHeartbeatTelemetryModule : ITelemetryModule, IDisposable
+    {
+        /// <summary>
+        /// Environment variables and the Application Insights heartbeat field names that accompany them.
+        /// </summary>
+        internal readonly KeyValuePair<string, string>[] WebHeartbeatPropertyNameEnvVarMap = new KeyValuePair<string, string>[]
+        {
+            new KeyValuePair<string, string>("appSrv_SiteName", "WEBSITE_SITE_NAME"),
+            new KeyValuePair<string, string>("appSrv_SiteName", "WEBSITE_SLOT_NAME"),
+            new KeyValuePair<string, string>("appSrv_wsStamp", "WEBSITE_HOME_STAMPNAME"),
+            new KeyValuePair<string, string>("appSrv_wsHost", "WEBSITE_HOSTNAME"),
+            new KeyValuePair<string, string>("appSrv_wsOwner", "WEBSITE_OWNER_NAME")
+        };
+
+        // for testing only: override the heartbeat manager
+        internal IHeartbeatPropertyManager HeartbeatManager;
+
+        private bool isInitialized = false;
+        
+        /// <summary>
+        /// Initializes a new instance of the<see cref="AppServicesHeartbeatTelemetryModule" /> class.
+        /// </summary>
+        public AppServicesHeartbeatTelemetryModule() : this(null)
+        {            
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AppServicesHeartbeatTelemetryModule" /> class. This is
+        /// internal, and allows for overriding the Heartbeat Property Manager to test this module with.
+        /// </summary>
+        /// <param name="hbeatPropManager">The heartbeat property manager to use when setting/updating env var values.</param>
+        internal AppServicesHeartbeatTelemetryModule(IHeartbeatPropertyManager hbeatPropManager)
+        {
+            this.HeartbeatManager = hbeatPropManager;
+        }
+
+        /// <summary>
+        /// Initialize the default heartbeat provider for Azure App Services. This module
+        /// looks for specific environment variables and sets them into the heartbeat 
+        /// properties for Application Insights, if they exist.
+        /// </summary>
+        /// <param name="configuration">Unused parameter.</param>
+        public void Initialize(TelemetryConfiguration configuration)
+        {
+            this.UpdateHeartbeatWithAppServiceEnvVarValues();
+            AppServiceEnvVarMonitor.Instance.MonitoredAppServiceEnvVarUpdatedEvent += this.UpdateHeartbeatWithAppServiceEnvVarValues;
+        }
+
+        /// <summary>
+        /// Signal the AppServicesHeartbeatTelemetryModule to update the values of the 
+        /// Environment variables we use in our heartbeat payload.
+        /// </summary>
+        public void UpdateHeartbeatWithAppServiceEnvVarValues()
+        {
+            try
+            {
+                var hbeatManager = this.GetHeartbeatPropertyManager();
```

Comment:
Right now, we are pretty much guaranteed that the HeartbeatPropertyManager is present (since it is the `DiagnosticsTelemetryModule` itself). And for the lifetime of the app it will either be there or it won't, no change. I think I will change this to cache it.

Reply:
Done.